### PR TITLE
refactor transformer

### PR DESCRIPTION
**Pull request by jshilong**
- This PR  is related to the basic operations of Vision Transformer. 

  1. Design a unified data-flow for the transformer.

  2. Design a `BaseTransformerLayer` as base-class of `TransformerEncoderlayer`  and `TransformerDecoderlayer`  in the vision transformer.  It contains several basic operations such as `attention`, `layer norm`, and `FFN`.  It can be built from `ConfigDict` and supports customization, for example, you can specify any number of `FFN or LN ` and use different kinds of `attention` by specifying a list of `ConfigDict` named `attn_cfgs`. It is worth mentioning that it supports `prenorm`.

  3. Design a ` BaseTransformerCoder` as base-class of `tranformer_coder(Encoder and Decoder)  ` in the vision transformer. Support customization such as specifying different kinds of `transformer_layer` in `tranformercoder`

#### 1 More unified data flow

 Design a  unified data flow for vision transformer and add `**kwargs` to adapt to other transformers (e.g., `deformable-detr`) .

```python
query (Tensor): Input query with the shape
    `(num_query, bs, embed_dims)`.
key (Tensor): The key tensor with shape
    `(num_key, bs, embed_dims)`.
value (Tensor): The value tensor with shape
    `(num_key, bs, embed_dims)`.
query_pos (Tensor): The positional encoding for `query`.
    Default: None.
key_pos (Tensor): The positional encoding for `key`. Default
    None.
attn_masks (list[Tensor], optional):  2D Tensor used in the calculation of corresponding attention.
    Defaults: None.
query_key_padding_mask (Tensor): ByteTensor for `query`, with
    shape [bs, num_query]. Only used in self-attention
key_padding_mask (Tensor): ByteTensor for `query`, with
    shape [bs, num_key].
```

#### 2 `BaseTransformerLayer`

It uses the unified data flow and can be initialized by basic arguments of all `TransformerLayer`  such  as 

```python
        attn_cfgs (list[`mmcv.ConfigDict`] | list[dict] | dict )):
            Configs for self_attention or cross_attention, the order
            should be consistent with it in `operation_order`. If it is
            a dict, it would be expanded to the number of attention in
            `operation_order`.
        feedforward_channels (int): The hidden dimension for FFNs.
        ffn_dropout (float): Probability of an element to be zeroed
            in ffn. Default 0.0.
        operation_order (tuple[str]): The execution order of operation
            in transformer. Such as ('selfattn', 'norm', 'ffn', 'norm').
            Support prenorm when you specifying first element as `norm`.
            Default：None
        act_cfg (dict): The activation config for FFNs.
        norm_cfg (dict): Config dict for normalization layer.
        ffn_num_fcs (int): The number of fully-connected layers in FFNs.
            Default：2.
```

The  `forward` function can be used in both `TransformerEncodeLayer` and `TransformerDecodeLayer` by specifying different `operation_order`.  It supports `pre_norm`  when you specifying the first operation as `norm` in `operation_order`.

#### 3 `BaseTransformerCoder`

It uses the unified data flow and can be initialized by basic arguments of all `BaseTransformerCoder`  such  as

```
        transformerlayer (list[obj:`mmcv.ConfigDict`] |
            obj:`mmcv.ConfigDict`): Config of transformerlayer
            in TransformerCoder. If it is obj:`mmcv.ConfigDict`,
             it would be repeated `num_layer` times to a
             list[`mmcv.ConfigDict`]. Default: None.
        num_layers (int): The number of `TransformerLayer`. Default: None.
```

# Usages

1.  `TransformerLayer`

you can build the `TransformerEncoderLayer`  by giving a` ConfigDict` which looks like 

```python
  
         codelayers=dict(
                type='BaseTransformerLayer',
                attn_cfgs=dict(
                    type='MultiheadAttention',
                    embed_dims=256,
                    num_heads=8,
                    dropout=0.1),
                feedforward_channels=2048,
                ffn_dropout=0.1,
                operation_order=('selfattn', 'norm', 'ffn', 'norm'),
            )
    # or more general,  you can specify different attentions in `TransformerLayer`.
            codelayers=dict(
                type='BaseTransformerLayer',
                attn_cfgs=[
                    dict(
                    	type='Attention_1',
                    	embed_dims=256,
                    	num_heads=8,
                    	dropout=0.1,
                    	other_args = xxx,
                        ...
                    ),
                    dict(
                    	type='Attention_2',
					    args=xxx,
                        ......
                    )     
                ],
                feedforward_channels=2048,
                ffn_dropout=0.1,
                operation_order=('selfattn', 'norm','selfattn', 'ffn', 'norm'),
            )
        
   # you can build the layer by
    from mmcv.cnn.bricks.transformer import build_transformerlayer
	build_transformerlayer(codelayers)
        
```

2. For `BaseTransformerCoder`， you can build it by

```python
            encoder=dict(
                type='DetrTransformerEncoder',
                num_layers=6,
                    transformerlayers=dict(
                        type='BaseTransformerLayer',
                        attn_cfgs=[dict(
                            type='MultiheadAttention',
                            embed_dims=256,
                            num_heads=8,
                            dropout=0.1)],
                        feedforward_channels=2048,
                        ffn_dropout=0.1,
                        operation_order=('self_attn', 'norm', 'ffn', 'norm'),
                    )
            ),

# or more general, you can specify different layers in the coder.
            encoder=dict(
                type='DetrTransformerEncoder',
                num_layers=2,
                    transformerlayers=dict(
                        type='BaseTransformerLayer',
                        attn_cfgs=[dict(
                            type='MultiheadAttention',
                            embed_dims=256,
                            num_heads=8,
                            dropout=0.1),
						dict(
                        	type='BaseTransformerLayer',
                        	attn_cfgs=[dict(
                            	type='xxx;ayer',
                            	args=xxx)
                         ],
                        feedforward_channels=2048,
                        ffn_dropout=0.1,
                        operation_order=('self_attn', 'norm', 'ffn', 'norm'),
                    )
            )
# you can build it by
from mmcv.cnn.bricks.transformer import build_transformercoder
build_transformerlayer(encoder)
```

## BC-breaking
The original DETR configs can't build models anymore, but you make compatibility with loading
original checkpoints.

### CODE REVIEW

File: configs/detr/detr_r50_8x2_150e_coco.py [hunk 1]
```diff
@@ -17,20 +17,36 @@
...
+                type='BaseTransformerLayer',
+                embed_dims=256,
+                attn_cfgs=dict(
+                    type='MultiheadAttention',
+                    embed_dims=256,
+                    num_heads=8,
+                    dropout=0.1),
+                feedforward_channels=2048,
+                ffn_dropout=0.1,
+                operation_order=('selfattn', 'norm', 'ffn', 'norm'),
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_8x2_150e_coco.py:33]: clean unnecessary comma

Comment (ZwwWayne) [inline configs/detr/detr_r50_8x2_150e_coco.py:33]: selfattn -> self_attn or attn?

File: mmdet/models/utils/transformer.py [hunk 2]
```diff
@@ -29,19 +34,21 @@ def __init__(self, embed_dims, num_heads, dropout=0.0):
...
+    def forward(
+        self,
+        query,
+        key=None,
+        value=None,
+        residual=None,
+        query_pos=None,
+        key_pos=None,
+        attn_mask=None,
+        key_padding_mask=None,
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:46]: clean comma

**Review by ZwwWayne — APPROVED** [review]

**Review by ZwwWayne — CHANGES_REQUESTED** [review]
Unit tests should be added to test the new modules, especially for those `base classes`

**Comment by ZwwWayne** [comment]
Overall, the design of refactoring looks good to me.

**Comment by ZwwWayne** [comment]
Resolve conflicts.

File: mmdet/models/dense_heads/detr_head.py [hunk 3]
```diff
@@ -607,14 +591,17 @@ def get_bboxes(self,
...
-                           img_shape,
-                           scale_factor,
-                           rescale=False):
+    def _get_bboxes_single(
+        self,
+        cls_score,
+        bbox_pred,
+        img_shape,
+        scale_factor,
+        rescale=False,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/detr_head.py:603]: clean comma

File: mmdet/models/utils/positional_encoding.py [hunk 4]
```diff
@@ -19,22 +18,27 @@ class SinePositionalEncoding(nn.Module):
...
-                 scale=2 * math.pi,
-                 eps=1e-6):
+    def __init__(
+        self,
+        num_feats,
+        temperature=10000,
+        normalize=False,
+        scale=2 * math.pi,
+        eps=1e-6,
+        offset=0.,
```
Comment (ZwwWayne) [inline mmdet/models/utils/positional_encoding.py:40]: clean comma.

File: mmdet/models/dense_heads/detr_head.py [hunk 5]
```diff
@@ -87,21 +73,9 @@ def __init__(self,
...
-        embed_dims = transformer['embed_dims']
-        assert num_feats * 2 == embed_dims, 'embed_dims should' \
-            f' be exactly 2 times of num_feats. Found {embed_dims}' \
-            f' and {num_feats}.'
-        assert test_cfg is not None and 'max_per_img' in test_cfg
-
+        self.bg_cls_weight = 0
         class_weight = loss_cls.get('class_weight', None)
-        if class_weight is not None:
+        if class_weight is not None and (self.__class__ is DETRHead):
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/detr_head.py:78]: can we use `isinstance(self, DETRHead)`?

File: mmdet/models/utils/transformer.py [hunk 6]
```diff
@@ -1,604 +1,154 @@
...
-                memory_attn_mask=None,
-                target_attn_mask=None,
-                memory_key_padding_mask=None,
-                target_key_padding_mask=None):
-        """Forward function for `TransformerDecoderLayer`.
+    def __init__(
+            self,
+            *args,
+            coder_norm_cfg=dict(type='LN'),
+            **kwargs,
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:73]: clean comma

File: mmdet/models/utils/transformer.py [hunk 7]
```diff
@@ -1,604 +1,154 @@
...
+    ):
+        super(DetrTransformerEncoder, self).__init__(*args, **kwargs)
+        if coder_norm_cfg is not None:
+            self.coder_norm = build_norm_layer(
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:77]: sefl.post_norm might be better than self.coder_norm?

File: mmdet/models/dense_heads/detr_head.py (hunk 5 above)
Comment (jshilong) [reply mmdet/models/dense_heads/detr_head.py:78]: We can't set class_weight in a subclass of `DETRHead`, for example, `DeformDETRhead`

File: configs/detr/README.md [hunk 8]
```diff
@@ -24,4 +24,4 @@ We provide the config files for DETR: [End-to-End Object Detection with Transfor
 
 | Backbone | Model | Lr schd | Mem (GB) | Inf time (fps) | box AP | Config | Download |
 |:------:|:--------:|:-------:|:--------:|:--------------:|:------:|:------:|:--------:|
-| R-50 | DETR  |150e |7.9|  | 40.1 | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/detr/detr_r50_8x2_150e_coco.py) | [model](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835-2c4b8974.pth) &#124; [log](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835.log.json) |
+| R-50 | DETR  |150e |7.9|  | 40.1 | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/detr/detr_r50_8x2_150e_coco.py) | [model](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20210417_195652-ab96448c.pth) &#124; [log](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20210417_195652-ab96448c.log.json) |
```
Comment (ZwwWayne) [inline configs/detr/README.md:27]: Add note to indicate that the best results may not be achieved in the last epoch

File: mmdet/models/dense_heads/detr_head.py [hunk 9]
```diff
@@ -23,45 +23,31 @@ class TransformerHead(AnchorFreeHead):
...
-                     num_decoder_layers=6,
-                     feedforward_channels=2048,
-                     dropout=0.1,
-                     act_cfg=dict(type='ReLU', inplace=True),
-                     norm_cfg=dict(type='LN'),
-                     num_fcs=2,
-                     pre_norm=False,
-                     return_intermediate_dec=True),
+                 num_query=100,
+                 reg_num_fcs=2,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/detr_head.py:49]: reg_num_fcs -> num_reg_fcs

File: mmdet/models/utils/transformer.py [hunk 10]
```diff
@@ -715,41 +208,28 @@ def forward(self, x, mask, query_embed, pos_embed):
...
+            query=target,
+            key=memory,
+            value=memory,
+            key_pos=pos_embed,
             query_pos=query_embed,
-            memory_attn_mask=None,
-            target_attn_mask=None,
-            memory_key_padding_mask=mask,
-            target_key_padding_mask=None)
+            key_padding_mask=mask,
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:224]: clean unnecessary comma

File: tests/test_models/test_dense_heads/test_detr_head.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        'img_shape': (s, s, 3),
+        'scale_factor': 1,
+        'pad_shape': (s, s, 3),
+        'batch_input_shape': (s, s)
+    }]
+    config = ConfigDict(
+        dict(
+            type='DETRHead',
+            num_classes=80,
+            in_channels=2048,
```
Comment (ZwwWayne) [inline tests/test_models/test_dense_heads/test_detr_head.py:20]: Use smaller channels for efficiency.

File: configs/detr/README.md [hunk 12]
```diff
@@ -24,4 +24,8 @@ We provide the config files for DETR: [End-to-End Object Detection with Transfor
 
 | Backbone | Model | Lr schd | Mem (GB) | Inf time (fps) | box AP | Config | Download |
 |:------:|:--------:|:-------:|:--------:|:--------------:|:------:|:------:|:--------:|
-| R-50 | DETR  |150e |7.9|  | 40.1 | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/detr/detr_r50_8x2_150e_coco.py) | [model](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835-2c4b8974.pth) &#124; [log](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835.log.json) |
+| R-50 | DETR  |150e |7.9|  | 40.1 | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/detr/detr_r50_8x2_150e_coco.py) | [model](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20210417_195652_c087b5d1.pth) &#124; [log](http://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20210417_195652_c087b5d1.log.json) |
```
Comment (ZwwWayne) [inline configs/detr/README.md:27]: Can we load the original checkpoint?

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**